man1/ftwhich.1
.\"
.\"     This is free software and distributed under the
.\"     terms of the Gnu Public License.
.\"
.TH ftwhich 15 "January 1999" "Search utilities"  \" -*- nroff -*-
.SH NAME
ftwhich \- fault tolerant search for a command name
.SH SYNOPSIS
.B ftwhich
[\-#hIp][\-t#] program_name
.SH DESCRIPTION
.B ftwhich 
is a fault tolerant version of the which(1) command.
.B ftwhich
searches for a given program in all directories included in your 
PATH environment variable and reports all files with a name that approximately
matches the given 
.B program_name.
.PP
.B ftwhich
achieves  fault tolerance  by calculating the  so called
Weighted Levenshtein Distance. The Levenshtein Distance
is defined as the  minimum  number of  character insertions,
deletions and replacements that transform 
a string 
.I A 
into a string 
.I B.
.PP
.I ftwhich
is similar to the
.I which 
command with the following differences:
.TP
\- 
ftwhich is by default NOT case sensitive
.TP
\- 
ftwhich is fault tolerant
.TP
\- 
Some shells have a build in
.I which
command that will also search aliases. 
.I ftwhich 
can naturaly not search for aliases as it does not know 
about alias definitions. 
.TP
\- 
.B ftwhich 
lists all files that approximatly match. The files first
shown take preference over files of the same name printed later as
they are from directories listed earlier in the PATH.
.TP
\- 
The level of fault tolerance can be adjusted by specifying the optional parameter 
.B tolerance.
A 
.B tolerance
of 0 specifies exact match.
.SH OPTIONS
.TP
.I \-h
Prints help/usage information.
.TP
.I \-I
Do case sensitive search (default is case in\-sensitive)
.TP
.I \-p
print the actual distance value in front of the found filename. This value
is equal to the number of insertions, deletions and replacements necessary
to transform the name of the found program into the search key.
.TP
.I \-# or \-t#
Set the fault tolerance level to #.
The fault tolerance level is an integer in the range 0-255.
It specifies the maximum number of errors permitted in finding 
the approximate match. The default tolerance is (strlen(searchpattern) - number of wildcards)/6 + 1
.PP
.TP 
.I program_name
The program file to search for.  
.RB '*'
and 
.RB '?' 
can be used as wildcards.  
.br
.RB '?'
denotes one single character.  
.br 
.RB '*'
denotes an arbitrary number of characters.
.PP
The last argument to ftwhich is not parsed for options as the program needs
at least one program_name argument. This means that 
.I ftwhich -x
will not complain about a wrong option but search for the program named -x.
.SH EXAMPLE
Search for all programs like gcc in your PATH:
.br
.I  ftwhich gcc
.br
This will e.g. find gcc or cc or CC ...
.PP
To find all files that start with any prefix and end in
.I config 
and differ in 2 letters from the word 
.I config:
.br
.I  ftwhich \-2
.RB '*config'
.PP
To find all files that exactly start with the prefix 
.I if:
.br
.I ftwhich \-0 
.RB 'if*' 
.PP
To find all clock programs:
.br
.I ftwhich \-0 '*clock*'
.PP
.SH BUGS
The wildcards 
.BR '?'
and 
.BR '*'
can not be escaped. These characters function always as
wildcards. This is however not a big problem since there is
normally hardly any command that has these characters in its name.
.SH AUTHOR
Guido Socher (guido@linuxfocus.org)
.SH SEE ALSO
whichman(1) ftff(1)



debian/changelog
whichman (2.4-8) unstable; urgency=medium

  * Add 11-CPPFLAGS-and-LDFLAGS.patch to pass the two variables
    to compiler.
  * Cleanup debian/rules, add support for cross-compiling, and
    enable LFS support.
  * Standards-Version: 3.9.5 (no changes).

 -- Robert Luberda <robert@debian.org>  Sat, 14 Dec 2013 21:35:44 +0100

whichman (2.4-7) unstable; urgency=low

  * Switch to debhelper v9 and tiny rules file.
  * Rename & refresh patches with gbp-pq import/export.
  * Fix lintian's `spelling-error-in-copyright'.
  * debian/rules: remove .pc/.dpkg-source-unapply file in rules.
    This fixes broken behaviour of dpkg-buildpackage (see Bug#649521).
  * debian/control:
    + Standards-Version: 3.9.2 (no changes);
    + add VCS fields;
    + sort dependency fields with wrap-and-sort.

 -- Robert Luberda <robert@debian.org>  Sun, 19 Feb 2012 16:05:46 +0100

whichman (2.4-6) unstable; urgency=low

  * Switch to the `3.0 (quilt)' format.
  * Bump to debhelper v8.
  * Standards-Version: 3.9.1 (no changes).
  * Split the previous 01_debian.patch into 3 patches: 00_Makefile.patch,
    01_documentation.patch, and 02_no_warn.patch.
  * debian/rules:
    + use dpkg-buildflags for getting CFLAGS/LDFLAGS (also added versioned
      build-dependency on dpkg-dev for this);
    + call dh_auto_{configure,build,clean} commands;
    + provide build-{arch,indep} targets.

 -- Robert Luberda <robert@debian.org>  Tue, 15 Feb 2011 22:29:25 +0100

whichman (2.4-5) unstable; urgency=low

  * Add copyright notice (lintian),

 -- Robert Luberda <robert@debian.org>  Sat, 19 Apr 2008 16:04:29 +0200

whichman (2.4-4) unstable; urgency=low

  * Use quilt to manage patches, move old debian patch to 01_debian.patch.
  * Build with debhelper v6.
  * 10_skip_duplicates.patch: Avoid duplicated files from `ftwhich' and
    `whichman' output (closes: #452068).
  * Standards-Version: 3.7.3 (no changes).
  * debian/control: Add Homepage URL.

 -- Robert Luberda <robert@debian.org>  Tue, 29 Jan 2008 22:39:17 +0100

whichman (2.4-3) unstable; urgency=low

  * Fix the `debian-rules-ignores-make-clean-error'.
    and manpage-has-errors-from-man' lintian errors.
  * Bump debhelper compat version to 6.
  * Standards-Version: 3.7.2 (no changes).

 -- Robert Luberda <robert@debian.org>  Mon, 15 Oct 2007 21:37:39 +0200

whichman (2.4-2) unstable; urgency=low

  * Fix typos in ftwhich(1) and ftff(1) man pages.
    (Closes: #306727,#306728).
  * Add debian/watch file.

 -- Robert Luberda <robert@debian.org>  Fri, 29 Apr 2005 19:06:43 +0200

whichman (2.4-1) unstable; urgency=low

  * New upstream version.

 -- Robert Luberda <robert@debian.org>  Tue, 20 Apr 2004 19:48:32 +0200

whichman (2.3-2) unstable; urgency=low

  * Opps, man pages went to /usr/share/man/man1/man1. Fixed.

 -- Robert Luberda <robert@debian.org>  Sun, 28 Mar 2004 23:15:17 +0200

whichman (2.3-1) unstable; urgency=low

  * New upstream version.
  * Standards-Version: 3.6.1 (no changes).

 -- Robert Luberda <robert@debian.org>  Wed, 24 Mar 2004 19:39:44 +0100

whichman (2.2-1) unstable; urgency=low

  * New upstream version.
  * debian/control: remove ending dot from the first line of the description
    field (lintian).
  * debian/copyright: update URLs.
  * Standards-Version: 3.5.8 (no changes).

 -- Robert Luberda <robert@debian.org>  Thu,  9 Jan 2003 00:01:13 +0100

whichman (2.0-3) unstable; urgency=low

  * Upgrade to Standards: 3.5.7
  * Build in debhelper in v4 mode.

 -- Robert Luberda <robert@debian.org>  Tue, 22 Oct 2002 22:20:23 +0200

whichman (2.0-2) unstable; urgency=low

  * Corrected spelling error in description (closes: #125481) and upstream
    README file. Thanks to Matt Ziemmerman.
  * Standards-Version: 3.5.6 (no changes).

 -- Robert Luberda <robert@debian.org>  Sat, 22 Dec 2001 22:22:00 +0100

whichman (2.0-1) unstable; urgency=low

  * New upstream version.
  * Fixed gcc 3.0 warnings.

 -- Robert Luberda <robert@debian.org>  Fri,  6 Jul 2001 22:49:21 +0200

whichman (1.9-2) unstable; urgency=low

  * Some fixes in manpages.
  * Updated debian/rules to the newest debhelper.
  * Standards-Version: 3.5.5.

 -- Robert Luberda <robert@debian.org>  Wed,  4 Jul 2001 22:58:46 +0200

whichman (1.9-1) unstable; urgency=low

  * Initial release (closes: #83624).
  * Use default Debian manpath if environment variable MANPATH
    is not set.

 -- Robert Luberda <robert@debian.org>  Mon, 29 Jan 2001 20:30:47 +0100


ftwhich.c
/* vim: set sw=8 ts=8 si et: */
/*
 * Author: Guido Socher. This program is distributed
 * under the terms of the Gnu Public License (GPL). 
 */
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <dirent.h>
#include <ctype.h>
#include <stdlib.h>
#include "levdist.h"

static void help();
static char **splitpath();
static int checkftype(char *path, char *filename);
static void matchfilename(char *path, char *key);
static int streq(const char *a,const char *b);

/*global for simplicity:*/
static int printdist=0;
static int casesensitive=0;
static int nomatchfound=1;
/*matchlimit indicates the tolerance level for Lev. Dist search
 *-2 not set, 
 * 0 = exact match with wildcards,
 * 1-255 max distance value*/
static int matchlimit=-2;


int main(int argc, char *argv[])
{
	char searchkey[MAXLEN +1];
	char **pathcomp;
        int optnum;
	char *c,*oarg;

        if (argc < 2 ) help();
        /*
         * After this while loop. argv[0] is the next argument if available
         *
         * options that take an argument write *c=0;break; others write
         * "c++;break;"
         *
         * This parser looks a bit complicated but it can take options
         * like -23 and -ac is the same as "-a -c"
         */
        optnum=argc;
        while (--optnum>0) {
                argv++;
                c = *argv; /* c is pointer to one argument/option */
                if (*c != '-'){
                        argv--;
                        break; /*stop at the first non option argument*/
                }
                c++;
                if (*c == '-' && *(c + 1)=='\0'){
                        /* you can write a double -- to stop option
                         * parsing. This is useful if you need to search
                         * for a string that is called "-something" */
                         --optnum;
                         break; /*stop option parsing*/
                }
                while (*c) {
                        switch (*c) {
                        case 'h':
                                help();
                                /*no break, help does not return*/
                        case 'p':
                                printdist=1;
                                c++;break;
                        case 'I':
                                casesensitive=1;
                                c++;break;
                        case 't':
                                /* you can write -t11 or -t 11*/
                                /* first the normal -t11 case: */
                                oarg=c+1;
                                /* then the -t 11 case: */
                                if(*(c + 1)=='\0'){
                                        if(--optnum) oarg=*++argv;
                                }
                                matchlimit = atoi(oarg);
                                if (matchlimit > 255 || matchlimit <0) matchlimit = 255;
                                *c=0;break;
                        default:
                                if (isdigit((int)*c)){
                                        if (matchlimit !=-1 ){
                                                /*ignore -e -t1 */
                                                matchlimit = atoi(c);
                                                if (matchlimit > 255 || matchlimit <0) matchlimit = 255;
                                        }
                                }else{
                                        fprintf(stderr,"ERROR: No such option. Use -h to get help.\n");
                                        exit(1);
                                }
                                *c=0;
                        }
                }
        }
        if (optnum<0) optnum=0; /* can be less then zero due to a option
                                 * with arg at last pos and missing arg*/
        /*
         *optnum  is now the number of argmunts left after
         *option parsing. Example: option a,c take no arg, option b has arg:
         *                         thisprogram -a -b xx -c d e
         *                         This results in:
         *                         optnum==2 and argv[1] points to d
         *                         argv[2] points to e
         */ 
	if (optnum != 1) {
		/* you must provide exactly one argument */
		help();
		/*help does not return */
	}
        strncpy(searchkey, argv[1],MAXLEN);
        if (matchlimit == -2){
                /*tolerance for Levenshtein Distance search: */
                matchlimit = stdtolerance(searchkey);
        }

	pathcomp = splitpath();
	while (*pathcomp) {
		matchfilename(*pathcomp, searchkey);
		pathcomp++;
	}
	return (nomatchfound);
}
/*__END OF MAIN__*/

/*
 * check that the file is a normal executable.
 * The abs. path to the file is "path/filename"
 * Return 1 if it is normal executable or a link to an normal executable
 */
#define MAXFULLPATHLEN 500
int checkftype(char *path, char *filename)
{
        struct stat stbuf;
        static char fullpath[MAXFULLPATHLEN];
        int i;
        /*construct the full path to the file:*/
        i=0;
        while(*path && i < MAXFULLPATHLEN-7){ /* -7 so, we have some room */
                fullpath[i]=*path;
                i++;path++;
        }
        if (fullpath[i-1]!='/'){
                fullpath[i]='/';
                i++;
        }
        while(*filename && i < MAXFULLPATHLEN-1){
                fullpath[i]=*filename;
                i++;filename++;
        }
        fullpath[i]='\0';
        /*check if file exists:*/
        if (stat(fullpath,&stbuf)!=0) return(0);
        /*not a regular file*/
        if ((stbuf.st_mode & S_IFREG)==0) return(0);
        /* test if executable for user or grp or others */
        if ((stbuf.st_mode & S_IXUSR) || (stbuf.st_mode & S_IXGRP)\
                ||(stbuf.st_mode & S_IXOTH)) return(1);
        return(0);
}
/*
 * match all entries in the directory variable path points to
 * against the key.
 * This function sets the global variable nomatchfound
 */
void matchfilename(char *path, char *key)
{
        int dist;
	DIR *dp;
	struct dirent *entry;
	if ((dp = opendir(path)) != NULL) {
                while ((entry = readdir(dp)) != NULL) {
                        /*ignore . or .. or .file: */
                        if (*(entry->d_name) == '.') continue; 
                        dist=levdist(entry->d_name, key, matchlimit,casesensitive);
                        if (dist != -1 && checkftype(path,entry->d_name)){
                                if (printdist) printf("%03d ",dist);
                                printf("%s/%s\n", path, entry->d_name);
                                nomatchfound = 0;
                        }
                }
                closedir(dp);
        }
}


/*
 * split PATH into substrings and return a pointer to 
 * a 2 dimmentional array with all the sub-strings of the
 * PATH. 
 */
char **splitpath()
{
	char *spath; /* search path */
	char *cspath;/* copy of search path */
	static char **pathcomp;
        int i=0;
        int notseen=1;
	char *start;
	int compind = 0;	/* index of *pathcomp[] */
	int complen = 1;	/* len of one component in the path + 16 */

	spath = (char *) getenv("PATH");
	if (spath == NULL || *spath == '\0') {
		fprintf(stderr,"Warning: PATH not defined, using /bin:/usr/bin\n");
                spath="/bin:/usr/bin";
	}
        /* make a copy that is static */
        cspath = (char *)malloc(strlen(spath) + 2);
        strcpy(cspath, spath);
	start = cspath;
        /* count the number of colons in the path to find how many
         * components we have in the path */
	while (*cspath) {
                if (*cspath == ':') i++;
		cspath++;
        }
        i+=3; /*some space in case we have corruped path below*/
        pathcomp=(char **)malloc(sizeof(char *)*i);
        /* split the path by replacing : with \0 */
	cspath = start;
        pathcomp[compind]=cspath;
	while (*cspath) {
		if (*cspath == ':') {
                        *cspath = '\0';
			if (complen > 1) { /*ignore zero length comp. e.g :: */
                                /* now check if it is a duplicate path 
                                 * component that we can just ignore: */
                                i=0;notseen=1;
                                while(notseen && i < compind){
                                        if(streq(pathcomp[compind],pathcomp[i])){
                                                notseen=0;
                                        }
                                        i++;
                                }
                                if (notseen) compind++;
			}
                        /* remove any tailing slashes but not if it is 
                         * only the root dir (just a "/") */
                        if (complen > 2 && *(cspath-1) == '/') *(cspath-1)='\0';
                        pathcomp[compind] = cspath+1;
                        complen = 0;
		}
		cspath++;
                complen++;
	}
        if (complen < 2){
		/*we have some corrupted path, this happens when you have a
		 *colon at the end of the path */
                pathcomp[compind] = NULL;   /*terminate */
        }else{
                /* now check if the last one is a duplicate path 
                 * component that we can just ignore: */
                i=0;notseen=1;
                while(notseen && i < compind){
                        if(streq(pathcomp[compind],pathcomp[i])){
                                notseen=0;
                        }
                        i++;
                }
                if (notseen) compind++;
                pathcomp[compind] = NULL;   /*terminate */
        }
	return (&pathcomp[0]);
}
/* return 1 if the 2 strings a and b are equal */
int streq(const char *a,const char *b){
	while(*a && *b){
		if(*a != *b) return(0);
		a++;b++;
	}
	/* both must be at \0 */
	if(*a || *b) return(0);
	return(1);
}
/*
 * help
 */
void help()
{
printf("ftwhich -- fault tolerant approximate search command for programs\n\
\n\
USAGE:  ftwhich [-#hIp][-t#] [--] program-name\n\
\n\
Supported wildcards: * -- any arbitrary number of character\n\
                     ? -- one character\n\
\n\
OPTIONS: -h  this help\n\
         -I  search case sensitive (default is case in-sensitive)\n\
         -p  print the actual distance (tolerance) value in front of the\n\
             file name.\n\
         -#  set fault tolerance level to # (integer in the range 0-255)\n\
             It specifies the maximum distance. This is the number of\n\
             errors permitted for finding the approximate match.\n\
         -t# same as -# for backward compatibility\n\
\n\
With no option specified search is fault tolerant using a tolerance\n\
level of: (string length of searchpattern - number of wildcards)/6 +1\n");
printf("%s\n",MYVERSION);
exit(0);
}
/*__END__*/
